---
title: "Introduction to Phylogenetics"
---

**Teaching: -- min** || **Exercises: -- min**

## Overview

:::::{.callout}

:::{.callout-important icon=false}
### Questions:
- List Question here
:::

:::{.callout-important icon=false}
### Learning Objectives:
- List learning objectives here
:::

:::{.callout-tip}
### Keypoints:
- List keypoints here
:::
:::::

## Multiple sequence alignment

## Phylogenetic tree inference

## Visualization of phylogenetic trees

There are many programs that can be used to visualise phylogenetic trees.

## Exercise

## Phylogenetic tree inference 

There are a number of different tools for phylogenetic inference via maximum-likelihood and some of the most popular tools used for phylogenetic inference are [`FastTree`](http://www.microbesonline.org/fasttree/), [`IQ-TREE`](http://www.iqtree.org/) and [`RAxML-NG`](https://github.com/amkozlov/raxml-ng). For this practical, we're going to use `IQ-TREE`.

First, create a directory for the `IQ-TREE` analysis and copy the alignment created by `nf-core/bactmap` into the new directory:

```bash
mkdir -p results/iqtree

cp results/bactmap/pseudogenomes/aligned_pseudogenomes.fas results/iqtree
```
### Variant site extraction with [`snp-sites`](https://github.com/sanger-pathogens/snp-sites)

Phylogenetic inference software such as `IQ-tree` typically takes as input an alignment of just the variant sites in a multiple sequence alignment.  So, before running `IQ-tree`, we need to extract the variant sites from the masked alignment.  To ensure that the branch lengths are correctly scaled, we also need to tell `IQ-tree` how many invariant or constant sites there are in our multiple sequence alignment.  To do both those things, we can use a tool called [`snp-sites`](https://github.com/sanger-pathogens/snp-sites). 

:::{.callout}
## Help
Do this to get the help information for `snp-sites`:
```bash
snp-sites -h
```
```bash
snp-sites: invalid option -- 'h'
Usage: snp-sites [-rmvpcbhV] [-o output_filename] <file>
This program finds snp sites from a multi FASTA alignment file.
 -r     output internal pseudo reference sequence
 -m     output a multi fasta alignment file (default)
 -v     output a VCF file
 -p     output a phylip file
 -o STR specify an output filename [STDOUT]
 -c     only output columns containing exclusively ACGT
 -C     only output count of constant sites (suitable for IQ-TREE -fconst) and nothing else
 -b     output monomorphic sites, used for BEAST
 -h     this help message
 -V     print version and exit
 <file> input alignment file which can optionally be gzipped

Example: creating files for BEAST
 snp-sites -cb -o outputfile.aln inputfile.aln

If you use this program, please cite:
"SNP-sites: rapid efficient extraction of SNPs from multi-FASTA alignments",
Andrew J. Page, Ben Taylor, Aidan J. Delaney, Jorge Soares, Torsten Seemann, Jacqueline A. Keane, Simon R. Harris,
Microbial Genomics 2(4), (2016). http://dx.doi.org/10.1099/mgen.0.000056
```
:::

#### Usage

The basic usage for `snp-sites` is:

```bash
snp-sites <ALIGNMENT> > <VARIANT SITE ALIGNMENT>
```

#### Run snp-sites

To extract the variant site alignment and count the number of constant sites in our alignment, run the following commands in the `results/iqtree`directory:

```bash
snp-sites \
    aligned_pseudogenomes.fas \
    -o aligned_pseudogenomes_snps.fas

snp-sites \
    -C aligned_pseudogenomes.fas \
    > constant_sites.txt
```

### Phylogenetic tree inference with [`IQ-TREE`](http://www.iqtree.org/)

:::{.callout}
## Help
Do this to get the help information for `IQ-TREE` (we've excluded much of what's printed as there's many, many options available for use with `IQ-TREE`):
```bash
iqtree -h
```
```bash
IQ-TREE multicore version 2.2.0.3 COVID-edition for Linux 64-bit built May 11 2022
Developed by Bui Quang Minh, James Barbetti, Nguyen Lam Tung,
Olga Chernomor, Heiko Schmidt, Dominik Schrempf, Michael Woodhams, Ly Trong Nhan.

Usage: iqtree [-s ALIGNMENT] [-p PARTITION] [-m MODEL] [-t TREE] ...

GENERAL OPTIONS:
  -h, --help           Print (more) help usages
  -s FILE[,...,FILE]   PHYLIP/FASTA/NEXUS/CLUSTAL/MSF alignment file(s)
  -s DIR               Directory of alignment files
  --seqtype STRING     BIN, DNA, AA, NT2AA, CODON, MORPH (default: auto-detect)
  -t FILE|PARS|RAND    Starting tree (default: 99 parsimony and BIONJ)
  -o TAX[,...,TAX]     Outgroup taxon (list) for writing .treefile
  --prefix STRING      Prefix for all output files (default: aln/partition)
  --seed NUM           Random seed number, normally used for debugging purpose
  --safe               Safe likelihood kernel to avoid numerical underflow
  --mem NUM[G|M|%]     Maximal RAM usage in GB | MB | %
  --runs NUM           Number of indepedent runs (default: 1)
  -v, --verbose        Verbose mode, printing more messages to screen
  -V, --version        Display version number
  --quiet              Quiet mode, suppress printing to screen (stdout)
  -fconst f1,...,fN    Add constant patterns into alignment (N=no. states)
  --epsilon NUM        Likelihood epsilon for parameter estimate (default 0.01)
  -T NUM|AUTO          No. cores/threads or AUTO-detect (default: 1)
  --threads-max NUM    Max number of threads for -T AUTO (default: all cores)
  --export-alisim-cmd  Export a command-line from the inferred tree and model params

  --> MORE OPTIONS
```

#### Usage

As indicated above, `IQ-TREE` has many different parameters available depending on what kind of tree you'd like to build including a comprehensive list of different nucleotide substitution models.  In practice, there only a few different options that need to be required to get a well-supported phylogentic tree using one of the the established nucleotide substitition models such as GTR or HKY.

```bash
iqtree \
    -fconst <NUMBER OF INVARIANT SITES FOR EACH NUCLEOTIDE>
    -s <INVARIANT SITE ALIGNMENT>
    -m <NUCLEOTIDE SUBSTITUTION MODEL>
    -bb <NUMBER OF ULTRAFAST BOOTSTRAPS TO PERFORM>
```

#### Run IQ-TREE

Now we can run `IQ-TREE` using the two files we just created with `snp-sites` as input:

```bash
iqtree \
    -fconst $(cat constant_sites.txt) \
    -s aligned_pseudogenomes_snps.fas \
    -nt auto \
    -ntmax 8 \
    -mem 8G \
    -m GTR \
    -bb 1000
```

The meaning of the options used is:

- `--fconst $(cat constant_sites.txt)` tells `IQ-tree` to extract the number of invariant sites from the file we created using `SNP-sites`
- `-s aligned_pseudogenomes_snps.fas` is the the variant site alignment
- `-nt auto` autodetects the number of available cores
- `-ntmax 8` is the maximal number of cores to use
- `-mem 8G` is the maximal RAM usage in GB
- `-m GTR` tells `IQ-tree` to use the General time reversible model of nucleotide substitution
- `-bb 1000` is the number of ultrafast bootstraps to run

We can look at the output directory (`results/iqtree`) to see the various  output files created by `IQ-tree`:

- `aligned_pseudogenomes_snps.fas.iqtree` is a text file containing a report of the IQ-Tree run, including a representation of the tree in text format.
- `aligned_pseudogenomes_snps.fas.treefile` is the estimated tree in NEWICK format. We can use this file with other programs, such as [`FigTree`](http://tree.bio.ed.ac.uk/software/figtree/), to visualise our tree. 
- `aligned_pseudogenomes_snps.fas.log` is the log file containing the messages that were also printed on the screen. 
- `aligned_pseudogenomes_snps.fas.bionj` is the initial tree estimated by neighbour joining (NEWICK format).
- `aligned_pseudogenomes_snps.fas.mldist` is the maximum likelihood distances between every pair of sequences.
- `aligned_pseudogenomes_snps.fas.ckp.gz` is this is a "checkpoint" file, which IQ-Tree uses to resume a run in case it was interrupted (e.g. if you are estimating very large trees and your job fails half-way through).

## Credit
Information on this page has been adapted and modified from the following source(s):
